modules/tinymce/src/core/main/ts/undo/Operations.ts

```diff
@@ -92,15 +99,36 @@ export const clear = (editor: Editor, undoManager: UndoManager, index: Index): v
   editor.dispatch('ClearUndos');
 };
 
-export const extra = (editor: Editor, undoManager: UndoManager, index: Index, callback1: () => void, callback2: () => void): void => {
-  if (undoManager.transact(callback1)) {
-    const bookmark = undoManager.data[index.get()].bookmark;
-    const lastLevel = undoManager.data[index.get() - 1];
-    Levels.applyToEditor(editor, lastLevel, true);
+export const extra: OperationsExtra = (editor: Editor, undoManager: UndoManager, locks: Locks, index: Index, callback1, callback2) => {
+  const isPromise = (entry: UndoLevel | null | Promise<UndoLevel | null>): entry is Promise<UndoLevel | null> => {
+    return Type.isNonNullable(entry) && 'then' in entry;
+  };
+
+  const extraContinue = (previousOutput: UndoLevel | null) => {
+    if (previousOutput) {
+      const extraEnd = () => {
+        undoManager.data[index.get() - 1].beforeBookmark = bookmark;
+      };
+
+      const bookmark = undoManager.data[index.get()].bookmark;
+      const lastLevel = undoManager.data[index.get() - 1];
+      Levels.applyToEditor(editor, lastLevel, true);
 
-    if (undoManager.transact(callback2)) {
-      undoManager.data[index.get() - 1].beforeBookmark = bookmark;
+      const continuedResult = transactAsync(undoManager, locks, callback2);
+
+      if (isPromise(continuedResult)) {
+        return continuedResult.then(extraEnd);
+      } else {
+        return extraEnd() as any;
```

Comment:
Why the any here it should be `void` or `Promise<void>` as the return type of the `extraContinue` function.

Reply:
Because typescript does not like it and would give 'void is not Promise<void>'-errors.

Reply:
Try adding a return type to the `extraContinue` any should never be needed.

Reply:
`Type 'void' is not assignable to type 'Promise<void>'`
Typescript really does not like the return type being void or something else.

Reply:
Yeah it's not hard to get to a point where TS is incapable of matching a perfectly valid type. Just go with `extraAsync` and avoid all the 'void or promise' stuff.